cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryCleanupService.java

```diff
@@ -0,0 +1,113 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The clean up service that cleans up cloned repositories periodically.
+ */
+public class RepositoryCleanupService extends AbstractScheduledService {
+  private final long cleanUpInterval;
+  private final long ttlInSeconds;
+  private ScheduledExecutorService executor;
+  private final Path repositoryCloneDirectory;
+
+  private static final Logger LOG = LoggerFactory.getLogger(RepositoryCleanupService.class);
+
+  @Inject
+  RepositoryCleanupService(CConfiguration cConf) {
+    this.cleanUpInterval = cConf.getLong(Constants.SourceControlManagement.REPOSITORY_CLEANUP_INTERVAL_SECONDS);
+    this.ttlInSeconds = cConf.getLong(Constants.SourceControlManagement.REPOSITORY_TTL_SECONDS);
+    this.repositoryCloneDirectory =
+        Paths.get(cConf.get(Constants.SourceControlManagement.GIT_REPOSITORIES_CLONE_DIRECTORY_PATH));
+  }
+
+  @Override
+  protected final ScheduledExecutorService executor() {
+    executor = Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("git-repository-cleanup"));
+    return executor;
+  }
+
+  @Override
+  protected void runOneIteration() {
+    deleteExpiredRepository();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    // Try right away if there's anything to cleanup
+    return Scheduler.newFixedRateSchedule(1, cleanUpInterval, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (executor != null) {
+      executor.shutdownNow();
+    }
+  }
+
+  /**
+   * The repositories are cloned under /<clone root>/namespace/<namesapce name>/
+   * see {@link RepositoryManager}
+   * @return List of repository directories
+   */
+  private List<File> getRepositoryDirectories(){
+    Path rootDirectoryPath = repositoryCloneDirectory.resolve(SourceControlConfig.cloneDirectoryPrefix);
+    List <File> namespaceDirectories = DirUtils.listFiles(rootDirectoryPath.toFile(), File::isDirectory);
+    return namespaceDirectories.stream().map(
+        d -> DirUtils.listFiles(d, File::isDirectory)
+    ).flatMap(List::stream).collect(Collectors.toList());
+  }
+
+  private void deleteExpiredRepository() {
+    try {
+    List<File> repositoryDirs = getRepositoryDirectories();
+    long currentTimeInSeconds = System.currentTimeMillis() / 1000;
+    for (File repositoryDir: repositoryDirs){
+        FileTime createTimeStamp = FileUtils.getFileCreationTime(repositoryDir.toPath());
+        if (createTimeStamp != null && (currentTimeInSeconds - createTimeStamp.toMillis() / 1000) > ttlInSeconds){
+          DirUtils.deleteDirectoryContents(repositoryDir);
```

Comment:
Maybe we should LOG here?

Reply:
It will mostly be unactionable hence I have not added any. I can add a debug log for edge case scenarios where we need to debug issues with wrong cleanup.